### PR TITLE
Removing the JsonPatchOperationFactory and just using TypeInfo annota…

### PR DESCRIPTION
**Pull request by hnguyen08**
…tion mixins for jackson deserialization/serialization

### CODE REVIEW

File: src/main/java/com/github/fge/jsonpatch/JsonPatch.java [hunk 1]
```diff
@@ -119,14 +120,15 @@ public JsonPatch(final List<JsonPatchOperation> operations)
      *
```
Comment (box-jhuffaker) [inline src/main/java/com/github/fge/jsonpatch/JsonPatch.java]: Do you need the constructor above?  You may I just wasn't sure.

**Comment by huggsboson** [comment]
As a client how do I create a JsonPatch which ExtendedJsonPatchOperationTypeInfoAnnotations?  Do I have to call readValue by myself?  It feels like StandardJsonPatchOperationTypeInfoAnnotations ExtendedJsonPatchOperationTypeInfoAnnotations should be somehow more closely linked to JsonPatch... all the way to making them extend JsonPatch (gasp, horror) and having an ExtendedJsonPatch and StandardJsonPatch...

also I'm somewhat tempted to lift ExtendedJsonPatch out of the core lib in a future commit and have it only live in our internal codebase while we evolve it.

File: src/main/java/com/github/fge/jsonpatch/ExtendedJsonPatch.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+public class ExtendedJsonPatch
+{
+    protected static final MessageBundle BUNDLE
+        = MessageBundles.getBundle(JsonPatchMessages.class);
+
+    private static final ObjectMapper mapper;
+
+    static
+    {
+        final ObjectMapper newMapper = JacksonUtils.newMapper();
```
Comment (huggsboson) [inline src/main/java/com/github/fge/jsonpatch/ExtendedJsonPatch.java]: mapper = JacksonUtils.newMapper();

File: src/main/java/com/github/fge/jsonpatch/ExtendedJsonPatch.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+                new NamedType(TestOperation.class, TestOperation.OPERATION_NAME),
+                new NamedType(OmitOperation.class, OmitOperation.OPERATION_NAME),
+                new NamedType(OmitOptionalOperation.class, OmitOptionalOperation.OPERATION_NAME)
+        );
+        mapper = newMapper;
+    }
+
+    public static ObjectReader getReader()
+    {
+        return mapper.reader();
```
Comment (huggsboson) [inline src/main/java/com/github/fge/jsonpatch/ExtendedJsonPatch.java]: I'm not sure if it saves anything if we cache the reader, like FGE did, worth thinking about.

File: src/main/java/com/github/fge/jsonpatch/JsonPatch.java [hunk 4]
```diff
@@ -89,12 +89,36 @@
...
+            new NamedType(CopyOperation.class, CopyOperation.OPERATION_NAME),
+            new NamedType(MoveOperation.class, MoveOperation.OPERATION_NAME),
+            new NamedType(RemoveOperation.class, RemoveOperation.OPERATION_NAME),
+            new NamedType(ReplaceOperation.class, ReplaceOperation.OPERATION_NAME),
+            new NamedType(TestOperation.class, TestOperation.OPERATION_NAME)
+        );
+        mapper = newMapper;
+    }
+
+    public static ObjectReader getReader()
```
Comment (huggsboson) [inline src/main/java/com/github/fge/jsonpatch/JsonPatch.java]: I'd probably just hide them.

File: src/main/java/com/github/fge/jsonpatch/ExtendedJsonPatch.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+import com.github.fge.jackson.JacksonUtils;
+import com.github.fge.jsonpatch.operation.*;
+import com.github.fge.msgsimple.bundle.MessageBundle;
+import com.github.fge.msgsimple.load.MessageBundles;
+
+import java.io.IOException;
+
+public class ExtendedJsonPatch
+{
+    protected static final MessageBundle BUNDLE
```
Comment (huggsboson) [inline src/main/java/com/github/fge/jsonpatch/ExtendedJsonPatch.java]: private

File: src/main/java/com/github/fge/jsonpatch/JsonPatch.java [hunk 6]
```diff
@@ -89,12 +89,36 @@
  * <p><b>IMPORTANT NOTE:</b> the JSON Patch is supposed to be VALID when the
  * constructor for this class ({@link JsonPatch#fromJson(JsonNode)} is used.</p>
  */
-public final class JsonPatch
+public class JsonPatch
     implements JsonSerializable
 {
-    private static final MessageBundle BUNDLE
+    protected static final MessageBundle BUNDLE
```
Comment (huggsboson) [inline src/main/java/com/github/fge/jsonpatch/JsonPatch.java]: private?

File: src/main/java/com/github/fge/jsonpatch/operation/JsonPatchOperation.java [hunk 7]
```diff
@@ -20,10 +20,13 @@
 package com.github.fge.jsonpatch.operation;
 
 import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonTypeInfo;
 import com.fasterxml.jackson.databind.JsonNode;
 import com.fasterxml.jackson.databind.JsonSerializable;
 import com.github.fge.jsonpatch.JsonPatchException;
 
+@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "op")
```
Comment (huggsboson) [inline src/main/java/com/github/fge/jsonpatch/operation/JsonPatchOperation.java]: I usually put annotations after doc comments

**Comment by huggsboson** [comment]
Sucks that this is what we have to do, but it looks like the only option.

File: src/main/java/com/github/fge/jsonpatch/ExtendedJsonPatch.java (hunk 5 above)
Comment (hnguyen08) [inline src/main/java/com/github/fge/jsonpatch/ExtendedJsonPatch.java]: Done

File: src/main/java/com/github/fge/jsonpatch/ExtendedJsonPatch.java (hunk 2 above)
Comment (hnguyen08) [inline src/main/java/com/github/fge/jsonpatch/ExtendedJsonPatch.java]: Done

File: src/main/java/com/github/fge/jsonpatch/ExtendedJsonPatch.java (hunk 3 above)
Comment (hnguyen08) [inline src/main/java/com/github/fge/jsonpatch/ExtendedJsonPatch.java]: Done

File: src/main/java/com/github/fge/jsonpatch/JsonPatch.java (hunk 6 above)
Comment (hnguyen08) [inline src/main/java/com/github/fge/jsonpatch/JsonPatch.java]: Done

File: src/main/java/com/github/fge/jsonpatch/JsonPatch.java (hunk 4 above)
Comment (hnguyen08) [inline src/main/java/com/github/fge/jsonpatch/JsonPatch.java]: We need to get the reader if we want to deserialize using Jackson not using fromJson (for example, if you want to deserialize an object that takes JsonPatch in its constructor)

File: src/main/java/com/github/fge/jsonpatch/operation/JsonPatchOperation.java (hunk 7 above)
Comment (hnguyen08) [inline src/main/java/com/github/fge/jsonpatch/operation/JsonPatchOperation.java]: Done

File: src/main/java/com/github/fge/jsonpatch/ExtendedJsonPatch.java [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+import com.fasterxml.jackson.databind.ObjectWriter;
+import com.fasterxml.jackson.databind.jsontype.NamedType;
+import com.github.fge.jackson.JacksonUtils;
+import com.github.fge.jsonpatch.operation.*;
+import com.github.fge.msgsimple.bundle.MessageBundle;
+import com.github.fge.msgsimple.load.MessageBundles;
+
+import java.io.IOException;
+
+public class ExtendedJsonPatch
```
Comment (huggsboson) [inline src/main/java/com/github/fge/jsonpatch/ExtendedJsonPatch.java]: We're really close.  I hate having the reader/writer hanging off the top level interfaces in the program.  Basically last thing I'd do is move ExtendedJsonPatch to ExtendedJsonPatchFactory, take the statics off so you can pass around the fromJson in tests.  Add your StandardJsonPatchFactory back with the mapper code from JsonPatch with its from json.  Then deprecate the static fromJson on JsonPatch and and add a static StandardJsonPatchFactory to it.  (circular dependency but oh well).  Then let's just fucking land it.  I stared at the jackson docs all night.

Comment (hnguyen08) [inline src/main/java/com/github/fge/jsonpatch/ExtendedJsonPatch.java]: I think I would have a JsonPatchFactory(mapper), then ExtendedJsonPatchFactory.create() create a JsonPatchFactory with the extended subtypes.  Makes it a bit easier to test and I think it still makes it easy for developers to select "standard" or "extended"

**Comment by huggsboson** [comment]
Merge it.

**Pull request merged**